vcap_utils/vcap_utils/backends/base_tensorrt.py

```diff
@@ -0,0 +1,278 @@
+        return resize.frame, resize
+
+    def allocate_buffers(self, batch_size: int = 1) -> \
+            Tuple[List[HostDeviceMem], List[HostDeviceMem], List[int], cuda.Stream]:
+        """Allocates host and device buffer for TRT engine inference.
+        Args:
+            batch_size: batch size for the input/output memory
+        Returns:
+            inputs [HostDeviceMem]: engine input memory
+            outputs [HostDeviceMem]: engine output memory
+            bindings [int]: buffer to device bindings
+            stream (cuda.Stream): cuda stream for engine inference synchronization
+        """
+        inputs = []
+        outputs = []
+        bindings = []
+        stream = cuda.Stream()
+        for binding in self.trt_engine:
+            size = trt.volume(self.trt_engine.get_binding_shape(binding)) * batch_size
+            dtype = trt.nptype(self.trt_engine.get_binding_dtype(binding))
+            # Allocate host and device buffers
+            host_mem = cuda.pagelocked_empty(size, dtype)
+            device_mem = cuda.mem_alloc(host_mem.nbytes)
+            # Append the device buffer to device bindings.
+            bindings.append(int(device_mem))
+            # Append to the appropriate list.
+            if self.trt_engine.binding_is_input(binding):
+                inputs.append(HostDeviceMem(host_mem, device_mem))
+            else:
+                outputs.append(HostDeviceMem(host_mem, device_mem))
+        return inputs, outputs, bindings, stream
+
+    def do_inference(self, bindings: List[int], inputs: List[HostDeviceMem], outputs: List[HostDeviceMem],
+                     stream: cuda.Stream, batch_size: int = 1) -> List[List[float]]:
+        # Transfer input data to the GPU.
+        self.ctx.push()
+        [cuda.memcpy_htod_async(inp.device, inp.host, stream) for inp in inputs]
+        # Run inference.
+        # todo: use async or sync api?
+        # According to https://docs.nvidia.com/deeplearning/tensorrt/best-practices/index.html#optimize-python
+        # the performance should be almost identical
+        self.context.execute(
+            batch_size=batch_size, bindings=bindings
+        )
+        # Transfer predictions back from the GPU.
+        [cuda.memcpy_dtoh_async(out.host, out.device, stream) for out in outputs]
+        # Synchronize the stream
+        stream.synchronize()
+        # Return only the host outputs.
+        batch_outputs = []
+        for out in outputs:
+            entire_out_array = np.array(out.host)
+            out_array_by_batch = np.split(entire_out_array, batch_size)
+            out_lists = [out_array.tolist() for out_array in out_array_by_batch]
+            batch_outputs.append(out_lists)
+        final_outputs = []
+        for i in range(len(batch_outputs[0])):
+            final_output = []
+            for batch_output in batch_outputs:
+                final_output.append(batch_output[i])
+            final_outputs.append(final_output)
```

Comment:
What's going on here? This section seems a bit confusing, and I'm wondering if it can't be simplified into a single for loop. I don't understand the need to have a `batch_outputs` and also a `final_outputs`.

Reply:
The output of batch prediction is a continuous array, for example for people detect, the output will be:
[[bboxes_img_1, bboxes_img_2,...,bboxes_img_n], [class_ids_1, class_ids_2...class_ids_n], [scores_1, scores_2...scores_n]]

This piece of code here will split the array so they will be formatted like this:
[[[bboxes_img_1], [class_ids_1], [scores_1]], [[bboxes_img_2], [class_ids_2], [scores_2]]...[[bboxes_img_n], [class_ids_n], [scores_n]]]

Reply:
In that case, you can replace lines 172-177 with 

```python
final_outputs = list(zip(*batch_outputs))
```

I tested it and it returns the same results.

Test code:

```python
def original_impl(batch_outputs):
    final_outputs = []
    for i in range(len(batch_outputs[0])):
        final_output = []
        for batch_output in batch_outputs:
            final_output.append(batch_output[i])
        final_outputs.append(final_output)
    return final_outputs


data = [["bbox1", "bbox2", "bbox3", "bbox4", "bbox5"],
        ["class1", "class2", "class3", "class4", "class5"],
        [1, 2, 3, 4, 5]]


def clean_impl(batch_outputs):
    return list(zip(*batch_outputs))


print(original_impl(data))
print(clean_impl(data))
```